File: py/structure/tree/search_tree/Treap.py
#!/usr/bin/env python
# -*- coding:gbk -*-

import random, sys

class Treap:
    def __init__(self):
        self.__root = None

    def isEmpty(self) -> bool:
        return not self.__root

    def add(self, key: int, value: int) -> None:
        self.__root = self.__addNode(self.__root, key, value)

    def erase(self, key: int) -> None:
        self.__root = self.__eraseNode(self.__root, key)

    def set(self, key: int, value: int) -> None:
        current = self.__root
        while current:
            if key == current.key:
                current.value = value
                return

            if key < current.key:
                current = current.left
            else:
                current = current.right

    def get(self, key: int) -> int:
        current = self.__root
        while current:
            if key == current.key:
                return current.value

            if key < current.key:
                current = current.left
            else:
                current = current.right
        return -sys.maxsize

    def clear(self) -> None:
        self.__root = None

    def __addNode(self, parent: '__TreeNode', key: int, value: int) -> '__TreeNode':
        if not parent:
            return self.__TreeNode(key, value)

        if key == parent.key:
            parent.value = value
            return parent

        if key < parent.key:
            parent.left = self.__addNode(parent.left, key, value)
            if parent.factor < parent.left.factor:
                parent = self.__rightRotation(parent)
        else:
            parent.right = self.__addNode(parent.right, key, value)
            if parent.factor < parent.right.factor:
                parent = self.__leftRotation(parent)
        return parent

    def __eraseNode(self, parent: '__TreeNode', key: int) -> '__TreeNode':
        if not parent:
            return None

        if key == parent.key:
            if parent.left and parent.right:
                if parent.left.factor > parent.right.factor:
                    parent = self.__rightRotation(parent)
                    parent.right = self.__eraseNode(parent.right, key)
                else:
                    parent = self.__leftRotation(parent)
                    parent.left = self.__eraseNode(parent.left, key)
            else:
                parent = parent.left if parent.left else parent.right
        elif key < parent.key:
            parent.left = self.__eraseNode(parent.left, key)
        else:
            parent.right = self.__eraseNode(parent.right, key)
        return parent

    def __leftRotation(self, current: '__TreeNode') -> '__TreeNode':
        child = current.right
        current.right, child.left = child.left, current
        return child

    def __rightRotation(self, current: '__TreeNode') -> '__TreeNode':
        child = current.left
        current.left, child.right = child.right, current
        return child

    class __TreeNode:
        def __init__(self, key: int, value: int):
            self.key, self.value, self.factor, self.left, self.right = key, value, random.randrange(0, 65536), None, None

File: java/structure/tree/search_tree/Treap.java
package structure.tree.search_tree;

import java.util.Random;

public class Treap {
    private TreeNode _root = null;

    public boolean isEmpty() {
        return null == _root;
    }

    public void add(int key, int value) {
        _root = addNode(_root, key, value);
    }

    public void erase(int key) {
        _root = eraseNode(_root, key);
    }

    public void set(int key, int value) {
        TreeNode current = _root;
        while (null != current) {
            if (key == current.key) {
                current.value = value;
                return;
            }

            if (key < current.key) {
                current = current.left;
            } else {
                current = current.right;
            }
        }
    }

    public int get(int key) {
        TreeNode current = _root;
        while (null != current) {
            if (key == current.key) {
                return current.value;
            }

            if (key < current.key) {
                current = current.left;
            } else {
                current = current.right;
            }
        }
        return Integer.MIN_VALUE;
    }

    public void clear() {
        _root = null;
    }

    private TreeNode addNode(TreeNode parent, int key, int value) {
        if (null == parent) {
            return new TreeNode(key, value);
        }

        if (key == parent.key) {
            parent.value = value;
            return parent;
        }

        if (key < parent.key) {
            parent.left = addNode(parent.left, key, value);
            if (parent.factor < parent.left.factor) {
                parent = rightRotation(parent);
            }
        } else {
            parent.right = addNode(parent.right, key, value);
            if (parent.factor < parent.right.factor) {
                parent = leftRotation(parent);
            }
        }
        return parent;
    }

    private TreeNode eraseNode(TreeNode parent, int key) {
        if (null == parent) {
            return parent;
        }

        if (key == parent.key) {
            if (null != parent.left && null != parent.right) {
                if (parent.left.factor > parent.right.factor) {
                    parent = rightRotation(parent);
                    parent.right = eraseNode(parent.right, key);
                } else {
                    parent = leftRotation(parent);
                    parent.left = eraseNode(parent.left, key);
                }
            } else {
                parent = (null != parent.left) ? parent.left : parent.right;
            }
        } else if (key < parent.key) {
            parent.left = eraseNode(parent.left, key);
        } else {
            parent.right = eraseNode(parent.right, key);
        }
        return parent;
    }

    private TreeNode leftRotation(TreeNode current) {
        TreeNode child = current.right;
        current.right = child.left;
        child.left = current;
        return child;
    }

    private TreeNode rightRotation(TreeNode current) {
        TreeNode child = current.left;
        current.left = child.right;
        child.right = current;
        return child;
    }

    private class TreeNode {
        public int key = 0;
        public int value = 0;
        public int factor = 1;
        public TreeNode left = null;
        public TreeNode right = null;

        public TreeNode(int key, int value) {
            this.key = key;
            this.value = value;
            this.factor = new Random().nextInt(65536);
        }
    }
}
